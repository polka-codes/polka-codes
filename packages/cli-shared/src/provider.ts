import { spawn } from 'node:child_process'
import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises'
import { dirname, normalize, resolve } from 'node:path'
import { input, select } from '@inquirer/prompts'
import type { ToolProvider } from '@polka-codes/core'
import ignore from 'ignore'
import { lookup } from 'mime-types'
import { checkRipgrep } from './utils/checkRipgrep'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command?: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
  summarizeOutput?: (stdout: string, stderr: string) => Promise<string | undefined>
  summaryThreshold?: number
}

export const getProvider = (options: ProviderOptions = {}): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  const memoryStore: Record<string, string> = {}
  const provider: ToolProvider = {
    listTopics: async (): Promise<string[]> => {
      return Object.keys(memoryStore)
    },
    read: async (topic: string): Promise<string | undefined> => {
      return memoryStore[topic]
    },
    append: async (topic: string, content: string): Promise<void> => {
      if (memoryStore[topic]) {
        memoryStore[topic] += content
      } else {
        memoryStore[topic] = content
      }
    },
    replace: async (topic: string, content: string): Promise<void> => {
      memoryStore[topic] = content
    },
    remove: async (topic: string): Promise<void> => {
      delete memoryStore[topic]
    },
    readFile: async (path: string, includeIgnored: boolean): Promise<string | undefined> => {
      if (!includeIgnored && ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      try {
        return await readFile(path, 'utf8')
      } catch (_e) {
        return undefined
      }
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    removeFile: async (path: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await unlink(path)
    },
    renameFile: async (sourcePath: string, targetPath: string): Promise<void> => {
      if (ig.ignores(sourcePath) || ig.ignores(targetPath)) {
        throw new Error(`Not allow to access file ${sourcePath} or ${targetPath}`)
      }
      return await rename(sourcePath, targetPath)
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number, includeIgnored: boolean): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, process.cwd(), options.excludeFiles, includeIgnored)
    },
    readBinaryFile: async (url: string) => {
      if (url.startsWith('file://')) {
        const filePath = decodeURIComponent(url.substring('file://'.length))
        const resolvedPath = normalize(resolve(process.cwd(), filePath))

        if (!resolvedPath.startsWith(process.cwd())) {
          throw new Error(`Access to file path "${filePath}" is restricted.`)
        }

        const data = await readFile(resolvedPath)
        const mediaType = lookup(resolvedPath) || 'application/octet-stream'

        return {
          base64Data: data.toString('base64'),
          mediaType,
        }
      }
      const response = await fetch(url)
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }
      const data = await response.arrayBuffer()
      const mediaType = lookup(url) || 'application/octet-stream'

      return {
        base64Data: Buffer.from(data).toString('base64'),
        mediaType,
      }
    },

    executeCommand: (
      command: string,
      _needApprove: boolean,
    ): Promise<{ stdout: string; stderr: string; exitCode: number; summary?: string }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command?.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', async (code) => {
          options.command?.onExit(code ?? 0)
          const totalLength = stdoutText.length + stderrText.length
          if (totalLength > (options.summaryThreshold ?? 5000) && options.summarizeOutput) {
            try {
              const summary = await options.summarizeOutput(stdoutText, stderrText)
              if (summary) {
                resolve({
                  summary,
                  stdout: stdoutText,
                  stderr: stderrText,
                  exitCode: code ?? 0,
                })
                return
              }
            } catch (_e) {
              console.error('Summarization failed:', _e)
            }
          }
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command?.onError(err)
          reject(err)
        })
      })
    },
    askFollowupQuestion: async (question: string, answerOptions: string[]): Promise<string> => {
      if (answerOptions.length === 0) {
        return await input({ message: question })
      }

      const otherMessage = 'Other (enter text)'
      answerOptions.push(otherMessage)
      const answer = await select({
        message: question,
        choices: answerOptions.map((option) => ({ name: option, value: option })),
      })

      if (answer === otherMessage) {
        return await input({ message: 'Enter your answer:' })
      }
      return answer
    },
    fetchUrl: async (url: string): Promise<string> => {
      const isRaw = url.startsWith('https://raw.githubusercontent.com/')

      const urlToFetch = isRaw ? url : `https://r.jina.ai/${url}`

      try {
        const response = await fetch(urlToFetch)
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`)
        }
        return await response.text()
      } catch (error) {
        console.error('Error fetching URL:', error)
        throw error
      }
    },
  }

  if (checkRipgrep()) {
    provider.searchFiles = async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, process.cwd(), options.excludeFiles)
    }
  } else {
    console.error(
      'Error: ripgrep (rg) is not installed. Search file tool is disabled. Please install it from https://github.com/BurntSushi/ripgrep#installation',
    )
  }

  return provider
}
