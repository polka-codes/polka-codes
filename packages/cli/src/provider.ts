import { spawn } from 'node:child_process'
import { mkdir, readFile, writeFile } from 'node:fs/promises'
import { dirname } from 'node:path'
import ignore from 'ignore'

import type { ToolProvider } from '@polka-codes/core'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
}

export const getProvider = (options: ProviderOptions): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  return {
    readFile: async (path: string): Promise<string> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await readFile(path, 'utf8')
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, dirname(path), options.excludeFiles)
    },
    searchFiles: async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, dirname(path), options.excludeFiles)
    },
    // listCodeDefinitionNames: async (path: string) => Promise<string[]> {},

    executeCommand: (command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', (code) => {
          options.command.onExit(code ?? 0)
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command.onError(err)
          reject(err)
        })
      })
    },
    // askFollowupQuestion: async (question: string, options: string[]) => Promise<string> {},
    // attemptCompletion: async (result: string) => Promise<string | undefined> {},
  }
}
