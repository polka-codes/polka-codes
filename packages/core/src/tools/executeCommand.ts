import { type FullToolInfo, PermissionLevel, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { CommandProvider } from './provider'
import { getBoolean, getString } from './utils'

export const toolInfo = {
  name: 'execute_command',
  description: `Request to execute a CLI command on the system. Use this when you need to perform system operations or run specific commands to accomplish any step in the user's task. You must tailor your command to the user's system and provide a clear explanation of what the command does. Prefer to execute complex CLI commands over creating executable scripts, as they are more flexible and easier to run. Commands will also be executed in the project root directory regardless of executed commands in previous tool uses.`,
  parameters: [
    {
      name: 'command',
      description:
        'The CLI command to execute. This should be valid for the current operating system. Ensure the command is properly formatted and does not contain any harmful instructions.',
      required: true,
      usageValue: 'Your command here',
    },
    {
      name: 'requires_approval',
      description: `A boolean indicating whether this command requires explicit user approval before execution in case the user has auto-approve mode enabled. Set to 'true' for potentially impactful operations like installing/uninstalling packages, deleting/overwriting files, system configuration changes, network operations, or any commands that could have unintended side effects. Set to 'false' for safe operations like reading files/directories, running development servers, building projects, and other non-destructive operations.`,
      required: false,
      usageValue: 'true or false',
    },
  ],
  examples: [
    {
      description: 'Request to execute a command',
      parameters: [
        {
          name: 'command',
          value: 'npm run dev',
        },
        {
          name: 'requires_approval',
          value: 'false',
        },
      ],
    },
  ],
  permissionLevel: PermissionLevel.Arbitrary,
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, CommandProvider> = async (provider, args) => {
  if (!provider.executeCommand) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to execute command. Abort.',
    }
  }

  const command = getString(args, 'command')
  const requiresApproval = getBoolean(args, 'requires_approval', false)
  const result = await provider.executeCommand?.(command, requiresApproval)

  const message = `<command>${command}</command>
<command_exit_code>${result.exitCode}</command_exit_code>
<command_stdout>
${result.stdout}
</command_stdout>
<command_stderr>
${result.stderr}
</command_stderr>`

  if (result.exitCode === 0) {
    return {
      type: ToolResponseType.Reply,
      message,
    }
  }
  return {
    type: ToolResponseType.Error,
    message,
  }
}

export const isAvailable = (provider: CommandProvider): boolean => {
  return !!provider.executeCommand
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfo
