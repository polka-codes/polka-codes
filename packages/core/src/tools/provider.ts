export type FilesystemProvider = {
  readFile?: (path: string) => Promise<string>
  writeFile?: (path: string, content: string) => Promise<void>
  removeFile?: (path: string) => Promise<void>
  listFiles?: (path: string, recursive: boolean, maxCount: number) => Promise<[string[], boolean]>
  searchFiles?: (path: string, regex: string, filePattern: string) => Promise<string[]>
  listCodeDefinitionNames?: (path: string) => Promise<string[]>
}

export type CommandProvider = {
  executeCommand?: (command: string, needApprove: boolean) => Promise<{ stdout: string; stderr: string; exitCode: number }>
}

export type InteractionProvider = {
  askFollowupQuestion?: (question: string, options: string[]) => Promise<string>
  attemptCompletion?: (result: string) => Promise<string | undefined>
}

export type ToolProvider = FilesystemProvider & CommandProvider & InteractionProvider

export class MockProvider implements ToolProvider {
  async readFile(path: string): Promise<string> {
    return 'mock content'
  }

  async writeFile(path: string, content: string): Promise<void> {
    return
  }

  async removeFile(path: string): Promise<void> {
    return
  }

  async listFiles(path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> {
    return [['mock-file.txt'], false]
  }

  async searchFiles(path: string, regex: string, filePattern: string): Promise<string[]> {
    return ['mock-file.txt']
  }

  async listCodeDefinitionNames(path: string): Promise<string[]> {
    return ['mockDefinition']
  }

  async executeCommand(command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return { stdout: 'mock output', stderr: '', exitCode: 0 }
  }

  async askFollowupQuestion(question: string, options?: string[]): Promise<string> {
    return 'mock answer'
  }

  async attemptCompletion(result: string): Promise<string | undefined> {
    return 'mock completion'
  }
}
