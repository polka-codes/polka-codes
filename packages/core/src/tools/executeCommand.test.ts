import { describe, expect, it, spyOn } from 'bun:test'
import executeCommand from './executeCommand'
import { MockProvider } from './provider'

describe('executeCommand', () => {
  it('should execute command successfully', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'executeCommand').mockResolvedValue({
      stdout: 'Command output',
      stderr: 'stderr',
      exitCode: 0,
    })

    const result = await executeCommand.handler(mockProvider, {
      command: 'echo test',
      requiresApproval: 'false',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.executeCommand).toHaveBeenCalledWith('echo test', false)
  })

  it('should handle command errors', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'executeCommand').mockRejectedValue(new Error('Command failed'))

    const result = executeCommand.handler(mockProvider, {
      command: 'invalid-command',
      requiresApproval: 'false',
    })

    // The handler now catches the error and returns a ToolResponseError
    const toolResponse = await result
    expect(toolResponse).toMatchSnapshot()
    expect(mockProvider.executeCommand).toHaveBeenCalledWith('invalid-command', false)
  })

  it('should handle command with non-zero exit code', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'executeCommand').mockResolvedValue({
      stdout: 'Command output',
      stderr: 'stderr',
      exitCode: 1,
    })

    const result = await executeCommand.handler(mockProvider, {
      command: 'invalid-command',
      requiresApproval: 'false',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.executeCommand).toHaveBeenCalledWith('invalid-command', false)
  })

  it('should handle approval required', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'executeCommand').mockResolvedValue({
      stdout: 'Approved command',
      stderr: 'stderr',
      exitCode: 0,
    })

    const result = await executeCommand.handler(mockProvider, {
      command: 'rm -rf /',
      requiresApproval: 'true',
    })

    expect(result).toMatchSnapshot
    expect(mockProvider.executeCommand).toHaveBeenCalledWith('rm -rf /', true)
  })
})
