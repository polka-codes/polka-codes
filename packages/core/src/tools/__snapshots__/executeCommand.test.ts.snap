// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`executeCommand should execute command successfully 1`] = `
{
  "message": 
"<command>echo test</command>
<command_exit_code>0</command_exit_code>
<command_stdout>
Command output
</command_stdout>
<command_stderr>
stderr
</command_stderr>"
,
  "type": "Reply",
}
`;

exports[`executeCommand should handle command errors 1`] = `[Error: Command failed]`;

exports[`executeCommand should handle command with non-zero exit code 1`] = `
{
  "message": 
"<command>invalid-command</command>
<command_exit_code>1</command_exit_code>
<command_stdout>
Command output
</command_stdout>
<command_stderr>
stderr
</command_stderr>"
,
  "type": "Error",
}
`;
